File: esw-dsl/src/main/scala/esw/dsl/TimeServiceDsl.scala
package esw.dsl

import java.time.Duration

import csw.time.core.models.{TMTTime, UTCTime}
import csw.time.scheduler.TimeServiceSchedulerFactory
import csw.time.scheduler.api.Cancellable

import scala.concurrent.ExecutionContext
import scala.concurrent.duration.FiniteDuration

trait TimeServiceDsl {
  private[esw] def timeServiceSchedulerFactory: TimeServiceSchedulerFactory

  def scheduleOnce(startTime: TMTTime)(task: => Unit)(implicit ec: ExecutionContext): Cancellable =
    timeServiceSchedulerFactory.make().scheduleOnce(startTime)(task)

  def schedulePeriodically(interval: FiniteDuration, startTime: TMTTime = UTCTime.now())(
      task: => Unit
  )(implicit ec: ExecutionContext): Cancellable =
    timeServiceSchedulerFactory.make().schedulePeriodically(startTime, Duration.ofNanos(interval.toNanos))(task)
}

File: esw-kt/script-dsl/src/main/kotlin/esw/ocs/dsl/highlevel/TimeServiceDsl.kt
package esw.ocs.dsl.highlevel

import csw.time.core.models.TAITime
import csw.time.core.models.TMTTime
import csw.time.core.models.UTCTime
import csw.time.scheduler.api.Cancellable
import csw.time.scheduler.api.TimeServiceScheduler
import java.util.concurrent.CompletionStage
import java.util.concurrent.TimeUnit
import kotlin.time.Duration
import kotlin.time.nanoseconds
import kotlin.time.toJavaDuration
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.future.future
import scala.concurrent.duration.FiniteDuration

interface TimeServiceDsl {
    val coroutineScope: CoroutineScope

    val timeServiceScheduler: TimeServiceScheduler

    suspend fun scheduleOnce(startTime: TMTTime, task: suspend () -> Unit): Cancellable =
        timeServiceScheduler.scheduleOnce(startTime, Runnable { task.toJavaFuture() })

    fun schedulePeriodically(
        startTime: TMTTime,
        interval: Duration,
        task: suspend () -> Unit
    ): Cancellable =
        timeServiceScheduler.schedulePeriodically(
            startTime,
            interval.toJavaDuration(),
            Runnable { task.toJavaFuture() })

    fun utcTimeNow(): UTCTime = UTCTime.now()

    fun taiTimeNow(): TAITime = TAITime.now()

    fun utcTimeAfter(duration: Duration): UTCTime =
        UTCTime.after(FiniteDuration(duration.toLongNanoseconds(), TimeUnit.NANOSECONDS))

    fun taiTimeAfter(duration: Duration): TAITime =
        TAITime.after(FiniteDuration(duration.toLongNanoseconds(), TimeUnit.NANOSECONDS))

    // fixme : move to appropriate place in clubed extension methods
    fun TMTTime.offsetFromNow(): Duration = durationFromNow().toNanos().nanoseconds

    private fun (suspend () -> Unit).toJavaFuture(): CompletionStage<Void> =
            coroutineScope.future { this@toJavaFuture }.thenAccept { }
}
